### PR TITLE
Support for the WASI ABI.

### PR DESCRIPTION
**Pull request by lachlansneff**
part of #297

### CODE REVIEW

File: lib/runtime-core/src/instance.rs [hunk 1]
```diff
@@ -63,7 +63,16 @@ impl Instance {
         // Initialize the vm::Ctx in-place after the backing
         // has been boxed.
         unsafe {
-            *inner.vmctx = vm::Ctx::new(&mut inner.backing, &mut inner.import_backing, &module)
+            *inner.vmctx = match imports.call_state_creator() {
+                Some((data, dtor)) => vm::Ctx::new_with_data(
```
Comment (bjfish) [inline lib/runtime-core/src/instance.rs]: `dtor`? maybe `destructor` or `finalizer`?

Comment (lachlansneff) [reply lib/runtime-core/src/instance.rs]: Yeah, either of those is fine. `dtor` was just short and it's common in the c++ world.

File: lib/wasi/src/syscalls/mod.rs [hunk 2]
```diff
@@ -0,0 +1,218 @@
+use crate::state::WasiState;
+use wasmer_runtime_core::{memory::Memory, vm::Ctx};
+
+#[allow(clippy::mut_from_ref)]
+fn get_wasi_state(ctx: &Ctx) -> &mut WasiState {
+    unsafe { &mut *(ctx.data as *mut WasiState) }
```
Comment (bjfish) [inline lib/wasi/src/syscalls/mod.rs]: One idea: we could differentiate between, ModuleState, InstanceState, FunctionState (if we add that).

File: lib/wasi/src/state.rs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+    pub inode_counter: Cell<u64>,
+}
+
+impl WasiFs {
+    pub fn new() -> Result<Self, String> {
+        zbox_init_env();
+        Ok(Self {
+            repo: RepoOpener::new()
+                .create(true)
+                .open("mem://foo", "")
```
Comment (xmclark) [inline lib/wasi/src/state.rs]: let's call it something else

File: lib/runtime-core/src/macros.rs [hunk 4]
```diff
@@ -57,6 +64,21 @@ macro_rules! imports {
...
+        };
+
+        let mut import_object = ImportObject::new_with_data($state_gen);
+
+        $({
+            let ns = $crate::__imports_internal!($ns);
+
+            import_object.register($ns_name, ns);
+        })*
+
```
Comment (syrusakbary) [inline lib/runtime-core/src/macros.rs]: What this does? What is `__imports_internal` ?

File: lib/wasi/src/syscalls/unix/macos.rs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+use crate::syscalls::types::*;
+use std::cell::Cell;
+
+pub fn platform_fd_pread(
```
Comment (xmclark) [inline lib/wasi/src/syscalls/unix/macos.rs]: Mark you know what to do

File: lib/wasi/src/syscalls/unix/linux.rs [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+                fd as i32,
+                iovec.as_mut_ptr(),
+                iovs_len as i32,
+                offset as i64,
+            ),
+            iovec,
+        )
+    };
+    nread.set(result as u32);
+    /*for (i, arr_cell) in iovs.iter().enumerate() {
```
Comment (xmclark) [inline lib/wasi/src/syscalls/unix/linux.rs]: 🗑

**Pull request merged**